kafka-connect-aws-s3/src/main/scala/io/lenses/streamreactor/connect/aws/s3/sink/S3WriterManager.scala

```diff
@@ -0,0 +1,169 @@
+
+    val newWriter = writer(topicPartitionOffset.toTopicPartition)
+    newWriter.write(struct, topicPartitionOffset)
+    newWriter.commitChecks()
+
+  }
+
+  /**
+    * Returns a writer that can write records for a particular topic and partition.
+    * The writer will create a file inside the given directory if there is no open writer.
+    */
+  def writer(topicPartition: TopicPartition): S3Writer = {
+    val bucketAndPrefix = bucketAndPrefixFn(topicPartition.topic)
+    writers.getOrElseUpdate(topicPartition, createWriter(bucketAndPrefix, topicPartition))
+  }
+
+  private def createWriter(bucketAndPrefix: BucketAndPrefix, topicPartition: TopicPartition): S3Writer = {
+
+    logger.debug(s"Creating new writer for bucketAndPrefix [$bucketAndPrefix]")
+
+    new S3WriterImpl(
+      bucketAndPrefix,
+      commitPolicyFn(topicPartition.topic),
+      formatWriterFn,
+      fileNamingStrategyFn(topicPartition.topic)
+    )
+  }
+
+  def preCommit(
+                 currentOffsets: Map[TopicPartition, OffsetAndMetadata]
+               ): Map[TopicPartition, OffsetAndMetadata] = {
+    logger.debug("Received call to S3WriterManager.preCommit")
+
+    currentOffsets
+      .collect {
+        case (topicPartition, offsetAndMetadata) =>
+          val writer: Option[S3Writer] = writers.get(topicPartition)
+          writer match {
+            case Some(writer: S3Writer) if writer.getCommittedOffset.isDefined =>
+              Some(topicPartition, createOffsetAndMetadata(offsetAndMetadata, writer))
+            case None => None
+          }
+      }.flatten.toMap
+  }
+
+  private def createOffsetAndMetadata(offsetAndMetadata: OffsetAndMetadata, writer: S3Writer) = {
+    new OffsetAndMetadata(
+      writer.getCommittedOffset.get.value,
+      offsetAndMetadata.leaderEpoch(),
+      offsetAndMetadata.metadata()
+    )
+  }
+}
+
+
+object S3WriterManager {
+  def from(config: S3Config)
+          (implicit storageInterface: StorageInterface): S3WriterManager = {
+
+    // TODO: make this configurable
+    val MinAllowedMultipartSize: Int = 5242880
+
+    val bucketAndPrefixFn: Topic => BucketAndPrefix = topic => config.bucketOptions.find(_.sourceTopic == topic.value)
+      .getOrElse(throw new ConnectException(s"No bucket config for $topic")).bucketAndPrefix
+
+    val commitPolicyFn: Topic => CommitPolicy = topic => config.bucketOptions.find(_.sourceTopic == topic.value) match {
+      case Some(bucketOptions) => bucketOptions.commitPolicy
+      case None => throw new IllegalArgumentException("Can't find commitPolicy in config")
+    }
+
+    val fileNamingStrategyFn: Topic => S3FileNamingStrategy = topic => config.bucketOptions
+      .find(_.sourceTopic == topic.value) match {
+      case Some(bucketOptions) => bucketOptions.fileNamingStrategy
+      case None => throw new IllegalArgumentException("Can't find fileNamingStrategy in config")
+    }
+
+    val minAllowedMultipartSizeFn: () => Int = () => {
+      // TODO : configuration
```

Comment:
What's left to do here?

Reply:
It would be good to have this as part of configuration.
However it didn't seem worth it until we support other cloud providers.
S3 has a single global limit, which is the limit defined in MinAllowedMultipartSize.